x-pack/plugin/index-lifecycle/src/main/java/org/elasticsearch/xpack/indexlifecycle/IndexLifecycleService.java

```diff
@@ -121,7 +123,9 @@ public void applyClusterState(ClusterChangedEvent event) {
         if (event.localNodeMaster()) { // only act if we are master, otherwise
                                        // keep idle until elected
             IndexLifecycleMetadata lifecycleMetadata = event.state().metaData().custom(IndexLifecycleMetadata.TYPE);
-            if (lifecycleMetadata != null && event.changedCustomMetaDataSet().contains(IndexLifecycleMetadata.TYPE)) {
+            if (lifecycleMetadata != null
+                    && (event.changedCustomMetaDataSet().contains(IndexLifecycleMetadata.TYPE) ||
+                        lifecycleMetadata.getPolicies().size() != policyRegistry.getLifecyclePolicyMap().size())) {
```

Comment:
> I'm not sure about this as a way to check if things have changed? what if a policy has been added and another has been removed? The number of policies will be equal but we should still run update().

In the case that policyRegistry is fairly up-to-date, and anything in the policymetadata changes, this will still call update because of the original changedcustomeMetaDataSet check. The size is only to reflect that policyRegistry was zeroed from being unelected or just recently elected to be master

> Is there a problem with always calling update() and just letting that method work out if things have changed?

I do not have a problem with this, I was just trying to preserve the
```
event.changedCustomMetaDataSet().contains(IndexLifecycleMetadata.TYPE)
```
check. If we do not have a problem with always updating, then this will simplify things for sure.

Reply:
I see. The size check is a bit of a cryptic way to check if the policy registry needs to be bootstrapped (which is essentially what its doing form your explanation?). Can we instead have a method on the policy registry which you can call to determine that it needs to be bootstrapped, or alternatively we can destroy the registry completely when we are no longer the master and recreate it here if its null?

Reply:
> I do not have a problem with this, I was just trying to preserve the
>```
>event.changedCustomMetaDataSet().contains(IndexLifecycleMetadata.TYPE)
>```

Given that we pass the whole cluster state to the registry we could also move this check inside the update method where we have more information and always call `update()` here.

Reply:
what do you think of going the always call update approach instead to simplify things even further?